huggingface_pipelines/pipeline.py

```diff
@@ -0,0 +1,64 @@
+from abc import ABC, abstractmethod
+from typing import Dict, Any
+import logging
+from datasets import Dataset
+from .pipeline_config import PipelineConfig
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+class Pipeline(ABC):
+    """
+    Abstract base class for different pipelines.
+    """
+
+    def __init__(self, config: PipelineConfig):
+        self.config = config
+        self.results = []
+
+    @abstractmethod
+    def process_batch(self, batch: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Processes a single batch of data and returns the updated batch.
+
+        Args:
+            batch (Dict[str, Any]): A batch of data.
+
+        Returns:
+            Dict[str, Any]: The updated batch.
+        """
+        pass
+
+    def __call__(self, dataset: Dataset) -> Dataset:
+        """
+        Processes the dataset and updates it.
+
+        Args:
+            dataset (Dataset): The dataset to process.
+
+        Returns:
+            Dataset: The updated dataset.
+        """
+        try:
+            logger.info("Starting to process dataset...")
+
+            if self.config.num_shards > 1:
+                dataset = dataset.shard(
+                    num_shards=self.config.num_shards, index=self.config.shard_id)
```

Comment:
to me sharding should be done only once (and typically at the beginning of the pipeline), I suppose that we dont need this generic functionality here (but more for dataloading). Yet keeping `take` makes sense for debugging maybe...